### PR TITLE
LZ Sampling Penalty

### PR DESCRIPTION
**Pull request by tginart**
## Motivation

Provide a reference implementation for the LZ penalty.

## Modifications

We introduce the LZ penalty, a sampling penalty specialized for reducing degenerate repetitions in auto-regressive language models without loss of capability. The penalty is based on the Lempel-Ziv sliding window universal data compression algorithm. We demonstrate the LZ penalty enables state-of-the-art open-source reasoning models to operate at with greedy (temperature zero) sampling without loss of capability and without any instances degenerate repetition.  Industry-standard repetition penalties incur degenerate repetition rates of up to $15$\% depending on the size of the model.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by tginart** [comment]
Preliminary results:

We apply the LZ penalty to $\texttt{Qwen/QwQ-32B}$ and $\texttt{deepseek-ai/DeepSeek-R1-Distill-Qwen-14B}$. We run GPQA and AIME benchmarks (averaging scores over 5 runs). We set a max token limit of $24k$. We fixed the $\texttt{top-p}$ to $0.95$ and the $\texttt{top-k}$ to $40$ for all runs (these are the recommended sampling parameters by the Qwen team). For all runs also using the LZ penalty, we fix the penalty strength $\alpha$ to $0.15$, the window size to $512$ and the buffer size to $32$. We found that this configuration of hyperparameters seemed to work well across both models and both datasets with minimal tuning required. We only tested other penalty strengths in an original sweep of 0.1, 0.2, and 0.3. We found that 0.2 was sometimes too high and 0.1 was sometimes too low. Thus 0.15 seemed to be a sweet spot. The window and buffer sizes were selected on intuition and did not seem to require changing.


Repetition & Capability:
image

image


Performance Benchmarking:
image

**Comment by tginart** [comment]
[Temp link to tech report draft](https://drive.google.com/file/d/1_tjl910IBHzriQD0o2o4p1vsWtigqFXn/view?usp=sharing)

**Comment by merrymercy** [comment]
Can you implement this with the custom logit processor?
https://github.com/sgl-project/sglang/blob/26c0f13126867f6fa8f14e872ba63c12171c51cf/test/srt/test_srt_endpoint.py#L310

The main branch will not accept these modifications to reduce the maintenance overhead.

**Comment by tginart** [comment]
Yes, sounds good!
